modules/javafx.graphics/src/main/java/com/sun/javafx/text/PrismTextLayout.java

```diff
@@ -421,66 +421,131 @@ public PathElement[] getCaretShape(int offset, boolean isLeading,
+                    for (int i = 0; i < runs.length; i++) {
+                        run = runs[i];
+                        if (x < run.getWidth()) {
+                            break;
+                        }
+                        if (i + 1 < runs.length) {
+                            if (runs[i + 1].isLinebreak()) {
+                                break;
+                            }
+                            x -= run.getWidth();
+                        }
                     }
                 }
             } else {
+                // This code branch is used to calculate hit info of Text node embedded in TextFlow.
                 for (int i = 0; i < lineIndex; i++) {
                     for (TextRun r: lines[i].runs) {
                         if (r.getTextSpan() != null && r.getStart() >= textRunStart && r.getTextSpan().getText().equals(text)) {
                             textWidthPrevLine += r.getLength();
                         }
                     }
                 }
-                int prevNodeLength = 0;
-                boolean isPrevNodeExcluded = false;
-                for (TextRun r: runs) {
-                    if (!r.getTextSpan().getText().equals(text) || (r.getStart() < textRunStart && r.getTextSpan().getText().equals(text))) {
-                        prevNodeLength += r.getWidth();
-                        continue;
+
+                BaseBounds textBounds = new BoxBounds();
+                if (isMirrored) {
+                    boolean isMultiRunText = false;
+                    for (TextRun r: runs) {
+                        if (r.getStart() != curRunStart && r.getTextSpan().getText().equals(text)) {
+                            isMultiRunText = true;
+                            break;
+                        }
                     }
-                    if (r.getTextSpan() != null && r.getTextSpan().getText().equals(text)) {
-                        BaseBounds textBounds = new BoxBounds();
-                        getBounds(r.getTextSpan(), textBounds);
-                        if (textBounds.getMinX() == 0 && !isPrevNodeExcluded) {
-                            x -= prevNodeLength;
-                            isPrevNodeExcluded = true;
+
+                    for (int i = 0; i < runs.length; i++) {
+                        run = runs[i];
+                        if (run.getStart() != curRunStart && run.getTextSpan().getText().equals(text) && x > run.getWidth()) {
+                            x -= run.getWidth();
+                            continue;
+                        }
+                        if (run.getTextSpan() != null && run.getTextSpan().getText().equals(text)) {
+                            if ((x > run.getWidth() && (!isMultiRunText || run.getStart() == curRunStart)) || textWidthPrevLine > 0) {
+                                getBounds(run.getTextSpan(), textBounds);
+                                x -= (runs[0].getLocation().x - textBounds.getMinX());
```

Comment:
suggestion: we are still in the for loop, so perhaps it makes sense to extract 
`(runs[0].getLocation().x - textBounds.getMinX());`
to a variable outside of the loop

Reply:
you are right: I missed the `break` in line 520